### PR TITLE
Add nuttx integration

### PR DESCRIPTION
**Pull request by jpbland1**
# Description

Adds NuttX integration files that allow a NuttX user to copy them into their apps/crypto folder and build with wolfssl after running the setup script

Relocate RTOS ports into new "rtos" folder.

Originally written by Kaleb Himes and Tim Weller and updated by John Bland.

# Testing

Tested with NuttX on the STM32H753

# Checklist

 - [*] added tests
 - [ ] updated/added doxygen
 - [ ] updated appropriate READMEs
 - [ ] Updated manual and documentation

### CODE REVIEW

**Review by dgarske — CHANGES_REQUESTED** [review]
Make sure you also add to include.am. Also we should consider putting this in a folder for RTOS or something not directly in root

File: nuttx/wolfssl/setup-wolfssl.sh [hunk 1]
```diff
@@ -0,0 +1,9 @@
+#!/bin/bash
+
+set -e # exit on any command failure
+if [ ! -d wolfssl ]; then
+    git clone https://github.com/wolfssl/wolfssl
+    git clone https://github.com/wolfssl/wolfssl-examples
+    cd wolfssl
+    patch -p 1 < ../wolfssl-nuttx.patch
```
Comment (dgarske) [inline nuttx/wolfssl/setup-wolfssl.sh:8]: Where is `wolfssl-nuttx.patch`?

**Review by dgarske — CHANGES_REQUESTED** [review]

File: nuttx/wolfssl/setup-wolfssl.sh (hunk 1 above)
Comment (jpbland1) [reply nuttx/wolfssl/setup-wolfssl.sh:8]: it is in nuttx/wolfssl. I actually want to make that patch conditionally compiled the way I did in tls13.c so we don't have to patch it at all

File: nuttx/include.am [hunk 2]
```diff
@@ -0,0 +1,10 @@
+EXTRA_DIST+= nuttx/include.am
```
Comment (dgarske) [inline nuttx/include.am:1]: You cannot include `nuttx/include.am`. Please add to `Makefile.am`.

Include the standard include.am header:

```
# vim:ft=automake
# included from Top Level Makefile.am
# All paths should be given relative to the root
```

The whole point of this is to include it with `make dist`.

File: nuttx/README.md [hunk 3]
```diff
@@ -0,0 +1,16 @@
+# wolfSSL NuttX integration
+
+This folder contains the instructions, scripts and build files neccessary to build NuttX with wolfSSL
```
Comment (dgarske) [inline nuttx/README.md:3]: Spelling: `neccessary` -> `necessary`

File: nuttx/wolfssl/setup-wolfssl.sh (hunk 1 above)
Comment (dgarske) [reply nuttx/wolfssl/setup-wolfssl.sh:8]: I don't see `wolfssl-nuttx.patch` in this PR?

**Review by dgarske — CHANGES_REQUESTED** [review]
Please find a way to move this into a new `rtos/nuttx` folder. I'd like to move `zephyr` in there too when this is merged.

File: nuttx/wolfssl/setup-wolfssl.sh (hunk 1 above)
Comment (jpbland1) [reply nuttx/wolfssl/setup-wolfssl.sh]: I've made it condionally compiled now, for reference the original patch file was this:
```
diff --git a/wolfcrypt/test/test.c b/wolfcrypt/test/test.c
index 4a670f391..40dd4f56a 100644
--- a/wolfcrypt/test/test.c
+++ b/wolfcrypt/test/test.c
@@ -573,7 +573,7 @@ WOLFSSL_TEST_SUBROUTINE int logging_test(void);
 #if !defined(NO_ASN) && !defined(NO_ASN_TIME)
 WOLFSSL_TEST_SUBROUTINE int time_test(void);
 #endif
-WOLFSSL_TEST_SUBROUTINE int mutex_test(void);
+WOLFSSL_TEST_SUBROUTINE int wolfcrypt_mutex_test(void);
 #if defined(USE_WOLFSSL_MEMORY) && !defined(FREERTOS)
 WOLFSSL_TEST_SUBROUTINE int memcb_test(void);
 #endif
@@ -1463,7 +1463,7 @@ options: [-s max_relative_stack_bytes] [-m max_relative_heap_memory_bytes]\n\
         TEST_PASS("time test passed!\n");
 #endif
 
-    if ( (ret = mutex_test()) != 0)
+    if ( (ret = wolfcrypt_mutex_test()) != 0)
         return err_sys("mutex    test failed!\n", ret);
     else
         TEST_PASS("mutex    test passed!\n");
@@ -42500,7 +42500,7 @@ WOLFSSL_TEST_SUBROUTINE int logging_test(void)
 }
 
 
-WOLFSSL_TEST_SUBROUTINE int mutex_test(void)
+WOLFSSL_TEST_SUBROUTINE int wolfcrypt_mutex_test(void)
 {
 #ifdef WOLFSSL_PTHREADS
     wolfSSL_Mutex m;
```

**Review by dgarske — CHANGES_REQUESTED** [review]
I think you are missing my comments here. I'd like this to move into `RTOS/nuttx`. And I'd like to also move zephyr into it (that can be a separate PR).

**Comment by jpbland1** [comment]
sorry I missed that comment, I was only reading the file comments

File: rtos/nuttx/wolfssl/README.md [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+    $ apt install -y kconfig-frontends
+    $ apt install -y gcc-arm-none-eabi binutils-arm-none-eabi
+    $ exit # exit sudo
+    ```
+3) Clone nuttx and nuttx-apps into working directory:
+    ```
+    $ git clone https://github.com/apache/nuttx.git nuttx
+    $ git clone https://github.com/apache/nuttx-apps apps
+    ```
+4) Copy this directory reanamed to "wolfssl" into the working directory applications:
```
Comment (dgarske) [inline rtos/nuttx/wolfssl/README.md:24]: Spelling `reanamed` -> `renamed`

File: rtos/nuttx/wolfssl/README.md [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+    ```
+    $ cd <path to where this archive was extracted>
+    $ cp -r wolfSSL-nuttX-app ~/nuttxspace/apps/crypto/wolfssl
+    ```
+5) Setup wolfSSL in preparation for the build:
+    ```
+    $ cd ~/nuttxspace/apps/crypto/wolfssl
+    $ ./setup-wolfssl.sh
```
Comment (dgarske) [inline rtos/nuttx/wolfssl/README.md:32]: Consider improving this. They just copied from wolfssl in the previous step.

File: rtos/nuttx/wolfssl/README.md [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+    ```
+> **NOTE:** There is a conflict with the function `mutex_test()` in `wolfcrypt/test/test.c`
```
Comment (dgarske) [inline rtos/nuttx/wolfssl/README.md:34]: Is this patch required still? Consider removing this comment.

File: rtos/nuttx/wolfssl/README.md [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+    NuttX Shell prompt (`nsh>`) should be displayed in the terminal.
+
+11) Run the wolfcrypt benchmark and/or test in the NuttX Shell:
+    ```
+    nsh> wolfcrypt_test
+    nsh> wolfcrypt_benchmark
+    ```
+## Notes
+- Developed using the following targets:
+    - STM NUCLEO-L552ZE-Q (Cortex-M33)
```
Comment (dgarske) [inline rtos/nuttx/wolfssl/README.md:76]: Didn't you also test on a NUCLEO-STM32H753?

File: rtos/nuttx/wolfssl/README.md [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+        - **wolfCrypt Benchmark application**
+        - **wolfCrypt Test application**
+    - Select Save from bottom menu, saving to `.config` file
+    - Exit configuration tool
+
+9) Build NuttX and wolfSSL:
+    ```
+    $ make
+    ```
+10) Open a serial terminal to the target, then load and run the NuttX image.  The
```
Comment (dgarske) [inline rtos/nuttx/wolfssl/README.md:66]: Did we miss a step for how to flash to the target?

File: rtos/nuttx/wolfssl/user_settings.h [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+#define SINGLE_THREADED
+#define WOLFSSL_SMALL_STACK
+
+/* Environment */
+#define NO_FILESYSTEM
+#define HAVE_STRINGS_H
+#define WOLF_C99
+
+/* Math */
+#ifdef CONFIG_ARCH_CHIP_STM32L552ZE
```
Comment (dgarske) [inline rtos/nuttx/wolfssl/user_settings.h:13]: Should there be something here for the STM32H753 also?

File: rtos/zephyr/include.am [hunk 10]
```diff
@@ -0,0 +1,45 @@
+# vim:ft=automake
```
Comment (dgarske) [inline rtos/zephyr/include.am:1]: Please fix `IDE/zephyr/README.md`

```
 # Zephyr
Zephyr Project Port has been moved to [wolfssl/zephyr](../../zephyr/README.md)
```

File: rtos/nuttx/wolfssl/README.md [hunk 11]
```diff
@@ -0,0 +1,78 @@
...
+
+6) Setup baseline NuttX configuration (board + NuttX Shell):
+    ```
+    $ cd ~/nuttxspace/nuttx
+    $ ./tools/configure.sh -l <board>:nsh
+    ```
+> **EXAMPLES:**
+>   - For NuttX Simulator: `$ ./tools/configure.sh sim:nsh`
+>   - For BL602 (RISC-V): `$ ./tools/configure.sh -l bl602evb:nsh`
+>   - For NUCLEO-L552ZE-Q (Cortex-M33): `$ ./tools/configure.sh -l nucleo-l552ze:nsh`
```
Comment (dgarske) [inline rtos/nuttx/wolfssl/README.md:46]: Where is the example for `NUCLEO-STM32H753`?

File: rtos/nuttx/wolfssl/README.md [hunk 12]
```diff
@@ -0,0 +1,78 @@
...
+
+1) Create working directory (e.g. ~/nuttxspace):
+    ```
+    $ cd ~
+    $ mkdir nuttxspace
+    ```
+2) Install dependencies:
+    ```
+    $ cd ~/nuttxspace
+    $ sudo su
```
Comment (dgarske) [inline rtos/nuttx/wolfssl/README.md:13]: Just use `sudo`, don't switch user. And remove the exit.

File: rtos/nuttx/wolfssl/Makefile [hunk 13]
```diff
@@ -0,0 +1,175 @@
+############################################################################
+# apps/crypto/wolfssl/Makefile
+#
+# Copyright (C) 2006-2020 wolfSSL Inc.
```
Comment (dgarske) [inline rtos/nuttx/wolfssl/Makefile:4]: 2023

File: rtos/nuttx/wolfssl/README.md [hunk 14]
```diff
@@ -0,0 +1,78 @@
...
+    - Enable and then select **wolfSSL applications**
+    - Enable applications:
+        - **wolfCrypt Benchmark application**
+        - **wolfCrypt Test application**
+    - Select Save from bottom menu, saving to `.config` file
+    - Exit configuration tool
+
+9) Build NuttX and wolfSSL:
+    ```
+    $ make
```
Comment (dgarske) [inline rtos/nuttx/wolfssl/README.md:64]: Example output:

```
$ make
Create version.h
LN: platform/board to /home/davidgarske/nuttxspace/apps/platform/dummy
Register: buttons
Register: wolfcrypt_benchmark
Register: wolfcrypt_test
Register: nsh
Register: sh
Register: stackmonitor_start
Register: stackmonitor_stop
Register: tee
CC:  wolfssl/wolfcrypt/src/random.c wolfssl/wolfcrypt/src/random.c:3781:10: warning: #warning "write a real random seed!!!!, just for testing now" [-Wcpp]
 3781 |         #warning "write a real random seed!!!!, just for testing now"
      |          ^~~~~~~
LD: nuttx
CP: nuttx.bin
```

File: rtos/nuttx/README.md [hunk 15]
```diff
@@ -0,0 +1,16 @@
+# wolfSSL NuttX integration
+
+This folder contains the instructions, scripts and build files necessary to build NuttX with wolfSSL
+
+## Installation
+```
+cp -R ./wolfssl <path-to-nuttx-apps>/crypto/
```
[the remark this answers is not in the document]
Comment (jpbland1) [reply rtos/nuttx/README.md:7]: ok I'll add a link to the original README, I'll change it to clone the local directory, I don't think symbolic links are going to work if they're using windows

File: rtos/nuttx/wolfssl/README.md (hunk 6 above)
Comment (jpbland1) [reply rtos/nuttx/wolfssl/README.md:34]: no patch is no longer required, I'll remove

File: RTOS/nuttx/wolfssl/user_settings.h [hunk 16]
```diff
@@ -0,0 +1,65 @@
...
+#define WOLFSSL_BENCHMARK_FIXED_UNITS_MB
+
+/* Development */
+/*#define DEBUG_WOLFSSL*/
+
+#define HAVE_TLS_EXTENSIONS
+#define HAVE_SUPPORTED_CURVES
+#define HAVE_ENCRYPT_THEN_MAC
+#define HAVE_EXTENDED_MASTER
+#define WOLFSSL_TLS13
```
Comment (dgarske) [inline RTOS/nuttx/wolfssl/user_settings.h:50]: This macro does not exist `WOLFSSL_TLS12`. Please remove (on by default unless `WOLFSSL_NO_TLS12` is set.

File: RTOS/nuttx/wolfssl/user_settings.h [hunk 17]
```diff
@@ -0,0 +1,65 @@
...
+    #define WOLFSSL_SP_MATH_ALL
+#endif
+
+/* Crypto */
+#define HAVE_ECC
+#define ECC_TIMING_RESISTANT
+#define WC_RSA_BLINDING
+#undef  RSA_LOW_MEM
+#define NO_MD4
+#define NO_DSA
```
Comment (dgarske) [inline RTOS/nuttx/wolfssl/user_settings.h:33]: Please add `NO_OLD_TLS`.

**Review by dgarske — CHANGES_REQUESTED** [review]

**Review by dgarske — APPROVED** [review]

File: RTOS/nuttx/wolfssl/README.md [hunk 18]
```diff
@@ -0,0 +1,202 @@
...
+    - Exit configuration tool
+
+    If you are using wolfSSL for TLS you should use the `netnsh` target and should enable an NTP or some for of system time keeping so that wolfSSL has the current date to check certificates. You will also need to set the right networking settings for NuttX to connect to the internet.
+9) Build NuttX and wolfSSL:
+    ```
+    $ make
+    ```
+    Example Output:
+    ```
+CC:  wolfssl/wolfcrypt/src/random.c /home/john/Documents/nuttxspace/apps/crypto/wolfssl/wolfssl/wolfcrypt/src/random.c:3781:10: warning: #warning "write a real random seed!!!!, just for testing now" [-Wcpp]
```
Comment (dgarske) [inline RTOS/nuttx/wolfssl/README.md:81]: This example make output is very poor. Either fix the warnings or don't include it...

**Review by dgarske — CHANGES_REQUESTED** [review]

**Review by dgarske — APPROVED** [review]

**Comment by dgarske** [comment]
Hi @maximevince, making sure you saw we moved the Zephyr directory into `RTOS/zephyr`. Let me know if you have any issues with that. Thanks, David Garske, wolfSSL

**Pull request merged**